snapshots/lvm/lvm.go

```diff
@@ -0,0 +1,135 @@
+// +build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package lvm
+
+import (
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+
+	"github.com/containerd/containerd/snapshots"
+	"github.com/pkg/errors"
+)
+
+func createLVMVolume(lvname string, vgname string, lvpoolname string, size string, fstype string, parent string, kind snapshots.Kind) (string, error) {
+	cmd := "lvcreate"
+	args := []string{}
+	out := ""
+	var err error
+
+	if parent != "" {
+		args = append(args, "--name", lvname, "--snapshot", vgname+"/"+parent)
+	} else {
+		// Create a new logical volume without a base snapshot
+		args = append(args, "--virtualsize", size, "--name", lvname, "--thin", vgname+"/"+lvpoolname)
+	}
+
+	if kind == snapshots.KindView {
+		args = append(args, "-pr")
+	}
+
+	//Let's go and create the volume
+	if out, err = runCommand(cmd, args); err != nil {
+		return out, errors.Wrap(err, "Unable to create volume")
+	}
+
+	if out, err = toggleactivateLV(vgname, lvname, true); err != nil {
+		return out, errors.Wrap(err, "Unable to activate thin volume")
```

Comment:
Do you encounter this kind of error?

Aborting.  LV mythinpool_tmeta is now incomplete and '--activationmode partial' was not specified.

or

Failed to suspend thin snapshot origin ...

Thanks.

Reply:
@renzhengeek i have not run into that error before? It seems like the error is happening within the metadata volume. Might i ask you for some more information on the system? What OS? How many thin volumes exist in that volumegroup?

Reply:
> @renzhengeek i have not run into that error before and it seems like the error is happening within the metadata volume. Might i ask you for some more information on the system? What OS? How many thin volumes exist in that volumegroup?

Hi, 

This error is likely caused by the underlaying disk missing out of sudden, in very short time, and come back again, retry to activate can succeed. 

So, it's not directly related to your snapshotter code:-) I asked in case others may see the same issue, if true, we may consider add retry logical to activate.

1k~2k thin LVs in the VG. By the way, how many in your side?

Reply:
I have pushed it up to 1.5K thin LVs. I have not had a scenario where the underlying disk disappears and re-appears shortly after. Re-tries do sound like a good idea. Let add that logic in.